### DISCUSSION TITLE
VRP - Finding a feasible solution given an initial route plus a new location

### DISCUSSION THREAD

**Discussion by tomliptrot** [google/or-tools#2970] [Routing (and legacy CP) questions]
I am trying to solve a Vehicle Routing Problem where new location are being added sequentially. My approach is:

1. Solve a Vehicle Routing Problem for a set of locations (with various constraints)
2. Add a new location to the problem
3. Solve the new problem using the solution from 1. as the starting point (ie not starting from scratch)

I can do this following the approach here: https://developers.google.com/optimization/routing/routing_tasks#setting-initial-routes-for-a-search

However, in order to make this work I must add the new location to the initial solution somehow.  This updated initial solution needs to be feasible or else it is rejected by `RoutingModel::DoRestoreAssignment()`.

In my case, finding a feasible starting solution can be difficult.

**Is there some way using or-tools, that I can efficiently find a 'good' single feasible solution as a starting point, given a previous solution and a new location?** 

I am doing this via python. Thanks in advance for a great piece of software.

**Comment by filippobistaffa** [comment]
Have you tried (initially) setting new locations as dropped nodes, and see if the solver can improve from that?

**Comment by tomliptrot** [comment]
Hi,

Thanks @filippobistaffa that is a good suggestion. Would that be done using `AddDisjunction`?

Tom

**Comment by filippobistaffa** [comment]
Yes, I would try that.

**Comment by superhqh** [comment]
> Hi,
> 
> Thanks @filippobistaffa that is a good suggestion. Would that be done using `AddDisjunction`?
> 
> Tom

I have tried that, but it didn't work.Have you solved your problem?

**Comment by filippobistaffa** [comment]
I am mixing the approach from [this example](https://developers.google.com/optimization/routing/penalties#add-the-capacity-constraints-and-penalties):
```
# Allow to drop nodes.
penalty = 1000
for node in range(1, len(data['distance_matrix'])):
    routing.AddDisjunction([manager.NodeToIndex(node)], penalty)
```
while also reading an initial assignment for (some of) the vehicles, as shown [here](https://developers.google.com/optimization/routing/routing_tasks#setting-initial-routes-for-a-search).
Make sure the initial routes you set correctly maps (in terms of node IDs) to your current instance.

**Comment by tomliptrot** [comment]
> > Hi,
> > Thanks @filippobistaffa that is a good suggestion. Would that be done using `AddDisjunction`?
> > Tom
> 
> I have tried that, but it didn't work.Have you solved your problem?

I did use that method and it did work. What goes wrong with yours @superhqh ?

**Comment by superhqh** [comment]
> > > Hi,
> > > Thanks @filippobistaffa that is a good suggestion. Would that be done using `AddDisjunction`?
> > > Tom
> > 
> > 
> > I have tried that, but it didn't work.Have you solved your problem?
> 
> I did use that method and it did work. What goes wrong with yours @superhqh ?

I am solving for VRPTW. There are 25 nodes and 1 vehicle. I first plan a feasible solution for the first 20 nodes. Then add the last 5 nodes in sequence as the initial solution, and set these 5 nodes as discardable nodes. Finally, I use ReadAssignmentFromRoutes to read the initial solution, and the result returns None, which means that the program thinks that the initial solution I provided is not a feasible solution.

**Reply by berkegokmen1** [reply]
Hi, any updates on this?
I wasn't able to produce a valid initial solution even though I added new nodes (pickup&delivery pair) to disjunction.

**Comment by zainabroo** [comment]
Hi.

Any progress on this? I'm trying an implementation very similar to that of @superhqh, using routes partially optimized by or-tools as an initial solution. When I read those routes via ReadAssignmentFromRoutes, None is returned. 

Also, to clarify: I've tried both including and excluding the start and end nodes from the list that defines the initial route for each of the vehicles, but no dice.  

Thanks, 
Zainab

**Comment by superhqh** [comment]
> Hi.
> 
> Any progress on this? I'm trying an implementation very similar to that of @superhqh, using routes partially optimized by or-tools as an initial solution. When I read those routes via ReadAssignmentFromRoutes, None is returned.
> 
> Also, to clarify: I've tried both including and excluding the start and end nodes from the list that defines the initial route for each of the vehicles, but no dice.
> 
> Thanks,
> Zainab
There is a solution i've tried, and it did successful. Make the partial routes you have confirmed as the initial solution, then make the additional nodes or whole nodes as optional use AddDisjunction.
